process/smartContract/process.go

```diff
@@ -1929,68 +1924,104 @@ func (sc *scProcessor) createAsyncCallBackSCRFromVMOutput(
-			result := createBaseSCR(outAcc, tx, txHash, 0)
-			sc.addVMOutputResultsToSCR(vmOutput, result)
-			return true, []data.TransactionHandler{result}
-		}
-
-		if !sc.flagDeploy.IsSet() {
-			result := createBaseSCR(outAcc, tx, txHash, 0)
-			result.Code = outAcc.Code
-			result.Value.Set(outAcc.BalanceDelta)
-			if result.Value.Cmp(zero) > 0 {
-				result.OriginalSender = tx.GetSndAddr()
-			}
+	return false, nil
+}
 
-			return false, []data.TransactionHandler{result}
-		}
+func (sc *scProcessor) preprocessOutTransferToSCR(
+	index int,
+	outputTransfer vmcommon.OutputTransfer,
+	outAcc *vmcommon.OutputAccount,
+	tx data.TransactionHandler,
+	txHash []byte,
+) *smartContractResult.SmartContractResult {
+	transferNonce := uint64(0)
+	if sc.flagIncrementSCRNonceInMultiTransfer.IsSet() {
+		transferNonce = uint64(index)
+	}
+	result := createBaseSCR(outAcc, tx, txHash, transferNonce)
 
-		return false, nil
+	if outputTransfer.Value != nil {
+		result.Value.Set(outputTransfer.Value)
+	}
+	result.Data = outputTransfer.Data
+	result.GasLimit = outputTransfer.GasLimit
+	result.CallType = outputTransfer.CallType
+	setOriginalTxHash(result, txHash, tx)
+	if result.Value.Cmp(zero) > 0 {
+		result.OriginalSender = tx.GetSndAddr()
+	}
+	return result
+}
+
+func (sc *scProcessor) createSmartContractResults(
+	vmOutput *vmcommon.VMOutput,
+	callType vmData.CallType,
+	outAcc *vmcommon.OutputAccount,
+	tx data.TransactionHandler,
+	txHash []byte,
+) (bool, []data.TransactionHandler) {
+
+	result := sc.createSCRFromStakingSC(outAcc, tx, txHash)
+	if !check.IfNil(result) {
+		return false, []data.TransactionHandler{result}
+	}
+
+	lenOutTransfers := len(outAcc.OutputTransfers)
+	if lenOutTransfers == 0 {
+		return sc.createSCRIfNoOutputTransfer(vmOutput, callType, outAcc, tx, txHash)
 	}
 
 	createdAsyncCallBack := false
-	var result *smartContractResult.SmartContractResult
 	scResults := make([]data.TransactionHandler, 0, len(outAcc.OutputTransfers))
 	for i, outputTransfer := range outAcc.OutputTransfers {
-		transferNonce := uint64(0)
-		if sc.flagIncrementSCRNonceInMultiTransfer.IsSet() {
-			transferNonce = uint64(i)
-		}
-		result = createBaseSCR(outAcc, tx, txHash, transferNonce)
-
-		if outputTransfer.Value != nil {
-			result.Value.Set(outputTransfer.Value)
-		}
-		result.Data = outputTransfer.Data
-		result.GasLimit = outputTransfer.GasLimit
-		result.CallType = outputTransfer.CallType
-		setOriginalTxHash(result, txHash, tx)
-		if result.Value.Cmp(zero) > 0 {
-			result.OriginalSender = tx.GetSndAddr()
-		}
+		result = sc.preprocessOutTransferToSCR(i, outputTransfer, outAcc, tx, txHash)
 
+		isCrossShard := sc.shardCoordinator.ComputeId(outAcc.Address) != sc.shardCoordinator.SelfId()
 		if result.CallType == vmData.AsynchronousCallBack {
-			createdAsyncCallBack = true
-			result.GasLimit, _ = core.SafeAddUint64(result.GasLimit, vmOutput.GasRemaining)
+			if !sc.flagCreatedCallBackCrossShardOnly.IsSet() || isCrossShard {
```

Comment:
originally, the L2021-L2023 were not conditioned by being only cross-shard. Is this correct for backwards compatibility reasons? What happened if we had an intra shard async call?

Reply:
if we had intra shard async call - the gasLimit would be added to this SCR and the SCR for sender with @ok would not be created.

Reply:
intra shard async calls and callbacks were already resolved by arwen. this is needed for cross shard only.

Reply:
ok